sycl/source/detail/image_accessor_util.cpp

```diff
@@ -29,12 +29,67 @@ cl_int4 getPixelCoordNearestFiltMode(cl_float4 Coord_uvw,
   cl_int height = ImgRange[1];
   cl_int depth = ImgRange[2];
   switch (SmplAddrMode) {
-  case addressing_mode::mirrored_repeat:
-    // TODO: Add the computations.
-    break;
-  case addressing_mode::repeat:
-    // TODO: Add the computations.
-    break;
+  case addressing_mode::mirrored_repeat: {
+    cl_float u1, v1, w1;
+    u1 = 2.0f * cl::sycl::rint(0.5f * u);
+    v1 = 2.0f * cl::sycl::rint(0.5f * v);
+    w1 = 2.0f * cl::sycl::rint(0.5f * w);
+
+    u1 = cl::sycl::fabs(u-u1);
+    v1 = cl::sycl::fabs(v-v1);
+    w1 = cl::sycl::fabs(w-w1);
+
+    u1 = u1 * width;
+    v1 = v1 * height;
+    w1 = w1 * depth;
+
+    i = (int)cl::sycl::floor(u1);
+    j = (int)cl::sycl::floor(v1);
+    k = (int)cl::sycl::floor(w1);
+
+    i = cl::sycl::min(i, (width-1));
+    j = cl::sycl::min(j, (height-1));
+    k = cl::sycl::min(k, (depth-1));
+    // Eg:
```

Comment:
I'd rather see an example before the code than after.
What do you think about the implementation of each case was in a separate function?

```
....
case addressing_mode::mirrored_repeat: {
  return mirrored_repeat_impl();
case addressing_mode::repeat:
  return repeat_impl(...);
case addressing_mode::clamp_to_edge:
  return clamp_to_edge_impl(...);
}
...
```

```
cl_int4 mirrored_repeat_impl(const cl_float4 Coord_uvw, const range<3> ImgRange) {
  auto calc = [](cl_int v, cl_int param) {
   cl_float temp = 2.0f * cl::sycl::rint(0.5f * v);
   temp = cl::sycl::fabs(v-temp);
   temp = temp * param;
   temp = cl::sycl::floor(temp)
   cl_int res = temp;
   res = cl::sycl::min(res, (param-1));
   return res;
  };
  
  const cl_int i = calc(*/u*/Coord_uvw.x(), /*width*/ImgRange[0]);
  const cl_int j = calc(*/v*/Coord_uvw.y(), /*height*/ImgRange[1]);
  const cl_int k = calc(*/w*/Coord_uvw.z(), /*depth*/ImgRange[2]);
  return {i, j, k, 0}
}
```

Such functions do one action, and are more easily covered by unit tests. Lambda calc also can be a function and tested(just give it a meaningful name.).

And instead of an example you will have a test :)

Reply:
I have made the code smaller by using vector operations. Let me know if you think I should populate separate functions.
Are you suggesting, I can write unit-tests in lit tests and call these functions individually to check if the function performs correctly?
I actually added the example as a comment, to help the code reviewer/reader understand what the code does. The code is tested in the imageaccessor_read_sampler.cpp test.
If you want I can remove these comments.